File: core/middleware.py
import secrets
import time
import uuid

from django.conf import settings as django_settings


class PathInfoNormalizerMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        if request.path != '/' and request.path.endswith('//'):
            request.path_info = request.path.rstrip('/')

        # Flask compatibility aliases used by Jinja templates.
        request.args = request.GET
        request.form = request.POST

        response = self.get_response(request)
        return response


class RequestIDMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        request.request_id = str(uuid.uuid4())
        request._start_ts = time.perf_counter()
        response = self.get_response(request)
        response['X-Request-ID'] = request.request_id
        # Only expose timing information in development
        if getattr(django_settings, 'DEBUG', False):
            duration_ms = int((time.perf_counter() - request._start_ts) * 1000)
            response['Server-Timing'] = f'app;dur={duration_ms}'
        return response


class CSPNonceMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        request.csp_nonce = secrets.token_urlsafe(16)
        response = self.get_response(request)
        nonce = getattr(request, 'csp_nonce', '')
        if nonce:
            csp = (
                "default-src 'self'; "
                f"script-src 'self' 'nonce-{nonce}' 'unsafe-eval' https://cdn.jsdelivr.net https://cdnjs.cloudflare.com https://cdn.tiny.cloud https://challenges.cloudflare.com; "
                "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com https://cdn.jsdelivr.net https://cdnjs.cloudflare.com https://cdn.tiny.cloud; "
                "img-src 'self' data: https:; "
                "font-src 'self' data: https://fonts.gstatic.com https://cdnjs.cloudflare.com; "
                "connect-src 'self' https://cdn.tiny.cloud https://challenges.cloudflare.com; "
                "frame-src 'self' https://challenges.cloudflare.com; "
                "frame-ancestors 'none'; base-uri 'self'; form-action 'self';"
            )
            if not getattr(django_settings, 'DEBUG', False):
                csp += " upgrade-insecure-requests;"
            response['Content-Security-Policy'] = csp
        return response


class SecurityHeadersMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        response = self.get_response(request)
        response.setdefault('X-Frame-Options', 'DENY')
        response.setdefault('X-Content-Type-Options', 'nosniff')
        response.setdefault('Referrer-Policy', 'strict-origin-when-cross-origin')
        response.setdefault('Cross-Origin-Resource-Policy', 'same-origin')
        response.setdefault('Cross-Origin-Opener-Policy', 'same-origin')
        response.setdefault('X-Permitted-Cross-Domain-Policies', 'none')
        response.setdefault('Origin-Agent-Cluster', '?1')
        response.setdefault(
            'Permissions-Policy',
            'camera=(), microphone=(), geolocation=(), payment=(), usb=(), '
            'magnetometer=(), gyroscope=(), accelerometer=()'
        )
        return response

File: righttechexperts-php/app/middleware.php
<?php
/**
 * Security Middleware
 * Replaces core/middleware.py (CSPNonceMiddleware, SecurityHeadersMiddleware, etc.)
 */

function run_middleware(): void
{
    // Generate CSP nonce
    $GLOBALS['__csp_nonce'] = rtrim(strtr(base64_encode(random_bytes(16)), '+/', '-_'), '=');
    $nonce = $GLOBALS['__csp_nonce'];

    // Generate request ID
    $requestId = sprintf(
        '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
        mt_rand(0, 0xffff),
        mt_rand(0, 0xffff),
        mt_rand(0, 0xffff),
        mt_rand(0, 0x0fff) | 0x4000,
        mt_rand(0, 0x3fff) | 0x8000,
        mt_rand(0, 0xffff),
        mt_rand(0, 0xffff),
        mt_rand(0, 0xffff)
    );

    // Set response headers
    header("X-Request-ID: {$requestId}");
    header('X-Frame-Options: DENY');
    header('X-Content-Type-Options: nosniff');
    header('Referrer-Policy: strict-origin-when-cross-origin');
    header('Cross-Origin-Resource-Policy: same-origin');
    header('Cross-Origin-Opener-Policy: same-origin');
    header('X-Permitted-Cross-Domain-Policies: none');
    header('Origin-Agent-Cluster: ?1');
    header('Permissions-Policy: camera=(), microphone=(), geolocation=(), payment=(), usb=(), magnetometer=(), gyroscope=(), accelerometer=()');

    // Content Security Policy
    $csp = "default-src 'self'; ";
    $csp .= "script-src 'self' 'nonce-{$nonce}' 'unsafe-eval' https://cdn.jsdelivr.net https://cdnjs.cloudflare.com https://cdn.tiny.cloud https://challenges.cloudflare.com; ";
    $csp .= "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com https://cdn.jsdelivr.net https://cdnjs.cloudflare.com https://cdn.tiny.cloud; ";
    $csp .= "img-src 'self' data: https:; ";
    $csp .= "font-src 'self' data: https://fonts.gstatic.com https://cdnjs.cloudflare.com; ";
    $csp .= "connect-src 'self' https://cdn.tiny.cloud https://challenges.cloudflare.com; ";
    $csp .= "frame-src 'self' https://challenges.cloudflare.com; ";
    $csp .= "frame-ancestors 'none'; base-uri 'self'; form-action 'self';";
    if (!APP_DEBUG) {
        $csp .= " upgrade-insecure-requests;";
    }
    header("Content-Security-Policy: {$csp}");
}
